libraries/AC_Avoidance/AC_Avoid.cpp

```diff
@@ -359,29 +359,54 @@ void AC_Avoid::adjust_velocity_z(float kP, float accel_cmss, float& climb_rate_c
         return;
     }
     
-    // do not adjust climb_rate if level or descending
-    if (climb_rate_cms <= 0.0f) {
+    // do not adjust climb_rate if level
+    if (is_zero(climb_rate_cms)) {
         return;
     }
 
+    const AP_AHRS &_ahrs = AP::ahrs();
     // limit acceleration
     const float accel_cmss_limited = MIN(accel_cmss, AC_AVOID_ACCEL_CMSS_MAX);
 
+    // if descending, adjust for minimum altitude as necessary
+    if (climb_rate_cms < 0.0f) {
+#if AP_FENCE_ENABLED
+        // calculate distance below fence
+        AC_Fence *fence = AP::fence();
+        if ((_enabled & AC_AVOID_STOP_AT_FENCE) > 0 && fence) {
+            // calculate distance from vehicle to safe altitude
+            float veh_alt;
+            _ahrs.get_relative_position_D_home(veh_alt);
+            if ((fence->get_enabled_fences() & AC_FENCE_TYPE_ALT_MIN) > 0) {
```

Comment:
I think these two could be combined?
```suggestion
        if ((_enabled & AC_AVOID_STOP_AT_FENCE) > 0 && fence && 
            (fence->get_enabled_fences() & AC_FENCE_TYPE_ALT_MIN) > 0) {
```

Reply:
Thanks fixed